### PR TITLE
Feature/enumerable methods

### PR DESCRIPTION
**Pull request by jubaan**
Enumerable methods project ready for first TSA reviews

### CODE REVIEW

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [x] my_each returns an Enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each.class # returns this error
   # NoMethodError:
       undefined method `call' for nil:NilClass
     # ./enumerable_methods.rb:6:in `block in my_each'
     # ./enumerable_methods.rb:5:in `times'
     # ./enumerable_methods.rb:5:in `my_each'
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_each_with_index returns an enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each_with_index.class # returns this error
    # NoMethodError:
       undefined method `call' for nil:NilClass
     # ./enumerable_methods.rb:13:in `block in my_each_with_index'
     # ./enumerable_methods.rb:12:in `times'
     # ./enumerable_methods.rb:12:in `my_each_with_index'
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_select returns an enumerator if no block is given
  ```
    array = [1,2,5,8,7,98,4]
    p array.my_select.class # return this errors
    # NoMethodError:
       undefined method `call' for nil:NilClass
     # ./enumerable_methods.rb:21:in `block in my_select'
     # ./enumerable_methods.rb:20:in `times'
     # ./enumerable_methods.rb:20:in `my_select'
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_all? returns true if the block never returns false or nil
  ```ruby
    array = [1,2,5,4,7]
    true_block = proc { |x| x < 10 }
    p array.my_all?(&true_block) == array.all?(&true_block) # false
  ```

* [x] my_all? when no block or argument is given returns true when none of the collection members are false or nil
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all? == array.all? # false
  ```
* [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all?(Integer) == array.all?(Integer) # false
  ```
* [x] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(/c/) == array.all?(/c/) # false
  ```
* [x] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(5) == array.all?(5) # false
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_any? returns true if the block ever returns a value other than false or nil
  ```ruby
    array = [1,2,5,4,7]
    true_block = proc {|num| num < 10 }
    p array.my_any?(&true_block) == array.any?(&true_block) # false
  ```

* [x] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
  ```ruby
    array = [nil, false, true, []]
    p array.my_any? == array.any? # false
  ```
* [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_any?(Integer) == array.any?(Integer) # false
  ```
* ( ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
  ```ruby
    array = %w[dog door rod blade]
    p array.my_any?(/c/) == array.any?(/c/) # false
  ```
* [x] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_any?(5) == array.any?(5) # false
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_none? returns true if the block never returns true for all elements
  ```ruby
    array = [1,2,5,4,7]
    false_block = proc { |x| x > 10 }
    p array.my_none?(&false_block) == array.none?(&false_block) # false
  ```

* [x] my_none? when no block or argument is given returns true only if none of the collection members is true
  ```ruby
    array = [1,2,5,4,7]
    p array.my_none? == array.none? # false
  ```
* [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_none?(Integer) == array.none?(Integer) # false
  ```
* [x] my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
  ```ruby
    array = %w[dog door rod blade]
    p array.my_none?(/c/) == array.none?(/c/) # false
  ```
* [x] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_none?(5) == array.none?(5) # false
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map #error
    # NoMethodError:
        undefined method `call' for nil:NilClass
      # ./enumerable_methods.rb:46:in `block in my_map'
      # ./enumerable_methods.rb:6:in `block in my_each'
      # ./enumerable_methods.rb:5:in `times'
      # ./enumerable_methods.rb:5:in `my_each'
      # ./enumerable_methods.rb:45:in `my_map'
  ```

* [x] my_map returns a new array with the results of running block once for every element in enum.
  ```ruby
    array = [1,2,5,4,7]
    block = proc { |num| num < 5 }
    p array.my_map?(&block) == array.map?(&block) # false
  ```

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_inject(:+) 
    # return this errors 
     # ArgumentError:
        tried to create Proc object without a block
      # ./enumerable_methods.rb:55:in `proc'
      # ./enumerable_methods.rb:55:in `block in my_inject'
      # ./enumerable_methods.rb:6:in `block in my_each'
      # ./enumerable_methods.rb:5:in `times'
      # ./enumerable_methods.rb:5:in `my_each'
      # ./enumerable_methods.rb:54:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Good start @jubaan :clap: :clap: However, some methods need further refactoring in order to get approval for the project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help refactor your methods :gear: 

- **[REQUIRED]** my_count method is inexistent, :pray: implement it.

- **[REQUIRED]** Enable the stickler so it shows if your code passes the linting checks

- **[REQUIRED]** Update the readme with info specific to your project

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
_**Ahmed Mahfoudh**_

**Comment by jubaan** [comment]
The stickler is making the suggestion of changing some stuff, but without it the code won't work. How can I disable the linter notifications?

**Review by whiz25 — CHANGES_REQUESTED** [review]
### Nice Progress @jubaan  :+1: 
I admire your resilience in doing your best to complete and merge this project. Your methods cover most of the test cases for this project. However, we have a few failing edge cases that we need to fix before we can merge.
- [x] Fix stickler errors
- [x] Check my comments for suggestions.

### Martin Afani

File: enumerable_methods.rb
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] calls the given block once for each element in self

File: enumerable_methods.rb
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex

File: enumerable_methods.rb
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable_methods.rb
Comment (whiz25) [inline enumerable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns the number of items in enum through enumeration
- [x] counts the number of items in enum that are equal to item if an argument is given
- [x] counts the number of elements yielding a true value if a block is given

File: enumerable_methods_yield.rb
Comment (whiz25) [inline enumerable_methods_yield.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] combines all elements of enum by applying a binary operation, specified by a block
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method

File: enumerable_methods.rb
Comment (stratospherique) [inline enumerable_methods.rb]: * [x] my_count returns the number of items in enum through enumeration
  ```ruby
    array = [1,2,5,4,7]
    p array.my_count == array.count # returns this error
      # NameError:
       undefined local variable or method `nilm' for #<Array:0x000056468bea49a8>
       Did you mean?  nil
                      nil?
     # ./enumerable_methods.rb:64:in `my_count'
  ```

* [x] my_count counts the number of elements yielding a true value if a block is given
```ruby
array = [1,2,8,9,7,5,6,3,11]
block = proc { |num| num < 5 }
array.my_count(&block) == array.count(&block) # returns this error
   # NameError:
       undefined local variable or method `nilm' for #<Array:0x000056468bec32e0>
       Did you mean?  nil
                      nil?
     # ./enumerable_methods.rb:64:in `my_count'
```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested changes :clap: However, "my_count" still needs further refactoring. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods

### That's it :relieved:  When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by Anna-Myzukina** [comment]
# Status [ APPROVED] image

I tested your methods and all cases passed.

### You can merge it. And don't forget to fill the project completion form.





*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*

**Pull request merged**
